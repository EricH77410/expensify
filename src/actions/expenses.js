
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => {
    return {
        type: 'ADD_EXPENSE',
        expense
    }
}

export const startAddExpense = (expenseData = {})=>{
   return (dispatch)=>{
    const {
        description='',
        note='',
        amount=0,
        createdAt=0
    } = expenseData
    const expense = {description, note, amount, createdAt}
    
    return database.ref('expenses').push(expense)
        .then((snap)=>{dispatch(addExpense({
            id: snap.key,
            ...expense
        }))})
   }
}
// REMOVE_EXPENSE
export const removeExpense = ({id}) => {
    return {
        type:'REMOVE_EXPENSE',
        id
    }
}
// ASYNC
export const startRemoveExpense = ({id})=>{
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove()
            .then(()=>{
                dispatch(removeExpense({id}))
            })
    }
}

// EDIT_EXPENSE
export const editExpense = (id,update) => {
    return {
        type: 'EDIT_EXPENSE',
        id,
        update
    }
}

// SET EXPENSES
export const setExpenses = (expenses) => {
    return {
        type: 'SET_EXPENSES',
        expenses
    }
}

export const startSetExpenses = () => {
    return (dispatch) => {
        const expenses = [];

        return database.ref('expenses').once('value').then((snap)=>{
            snap.forEach((childSnap)=>{
                expenses.push({
                    id:childSnap.key,
                    ...childSnap.val()
                })
            })
        }).then(()=>{
            dispatch(setExpenses(expenses));
        })
    }
}


