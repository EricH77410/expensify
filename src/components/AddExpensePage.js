import React from 'react'

const AddExpensePage = () => {
    return (
        <h3>Hello from new</h3>
    )
}

export default AddExpensePage