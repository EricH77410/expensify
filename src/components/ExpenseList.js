import React from 'react';
import { connect } from 'react-redux';

import ExpenseItem from './ExpenseItem'
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (

    <div>
        {
            props.expenses.length === 0 ? (
                <p>Add a new expense to begin ...</p>
            ) : (
                props.expenses.map((e)=>{
                    return <ExpenseItem {...e} key={e.id}/>
                })
            )
        }

       
        
    </div>
)
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses,state.filters),
        filters: state.filters
    }
}

export default connect(mapStateToProps)(ExpenseList);