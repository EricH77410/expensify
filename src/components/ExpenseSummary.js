import React from 'react';
import { connect } from 'react-redux';
import getExpensesTotal from '../selectors/expenses-total';
import getVisibleExpenses from '../selectors/expenses'
import numeral from 'numeral';

export const ExpenseSummary = ({expensesCount, expensesTotal}) => {
    const word = expensesCount > 1 ? 'expenses' : 'expense'
    return (
        <div>
            <h3>Expense Summary</h3>
            <p>Viewing {expensesCount} {word}</p>
            <p>Total : {numeral(expensesTotal/100).format('$0,0.00')}</p>
        </div>
    )
}

const mapStateToProps = (state)=>{
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters)
    return {
        expensesCount: visibleExpenses.length,
        expensesTotal: getExpensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpenseSummary);