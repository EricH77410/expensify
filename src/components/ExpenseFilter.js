import React from 'react';
import { connect } from 'react-redux'
import { DateRangePicker } from 'react-dates'

import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters'

export class ExpenseFilter extends React.Component {
    state = {
        calendarFocused:null,
    }
    onFocusChange = (calendarFocused) => {
        this.setState(()=>({calendarFocused}))
    }
    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    onSortChange = (e) => {
        e.target.value === 'date' ? this.props.sortByDate() : this.props.sortByAmount()
    }
    render() {
        return (
            <div className="input-filter">
                <input 
                type="text" 
                value={this.props.filters.text} 
                onChange={this.onTextChange}/>
                <select 
                value={this.props.filters.sortBy}
                onChange={this.onSortChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                    showClearDates={true}
                />
            </div>
        )
    }
    
};
const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        setTextFilter: (txt)=>dispatch(setTextFilter(txt)),
        sortByDate: () => dispatch(sortByDate()),
        sortByAmount: ()=>dispatch(sortByAmount()),
        setStartDate: (date)=>dispatch(setStartDate(date)),
        setEndDate: (date)=>dispatch(setEndDate(date))
    }
}
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseFilter);