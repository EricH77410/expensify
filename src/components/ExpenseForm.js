import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';

import { addExpense } from '../actions/expenses';

export default class ExpenseForm extends React.Component{
    state = {
        description:'',
        amount:0,
        note:'',
        createdAt: moment(),
        calendarFocused: false
    }
    descriptionChange=(e)=>{
        const description = e.target.value;
        this.setState(()=>({description}))
    }
    noteChange=(e)=>{
        const note = e.target.value;
        this.setState(()=>({note}));
    }
    amountChange=(e)=>{
        const amount = e.target.value;
        if(amount.match(/^\d*(\.\d{0,2})?$/)){
            this.setState(()=>({amount}));
        }
    }
    dateChange=(createdAt)=>{
        this.setState(()=>({createdAt}))
    }
    focusChange=({focused})=>{
        this.setState(()=>({calendarFocused: focused}));
    }
    render(){
        return (
            <div>
                <form>
                   <input 
                   type="text" 
                   placeholder="description" 
                   value = {this.state.description}
                   onChange={this.descriptionChange}
                   autoFocus/>
                   <input 
                   type="text" 
                   placeholder="Amount"
                   value={this.state.amount}
                   onChange={this.amountChange}
                   />
                   <SingleDatePicker
                   date={this.state.createdAt}
                   onDateChange={this.dateChange}
                   focused={this.state.calendarFocused}
                   onFocusChange={this.focusChange}
                   numberOfMonths={1}
                   isOutsideRange={()=> false}
                    />
                   <textarea 
                   placeholder="Add a note..."
                   value={this.state.note}
                   onChange={this.noteChange}
                   >
                   </textarea>
                   <button>Add Expense</button>
                </form>
            </div>
        )
    }
}