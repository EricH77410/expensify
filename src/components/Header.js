import React from 'react'
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom'
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout }) => (
    <div>
        <h1>Expensify App</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink>
        <NavLink to="/create" activeClassName="is-active">Create</NavLink>
        <button onClick={startLogout}>Logout</button>
    </div>
)

const mapDispatchToProps = (dispatch) => {
    return {
        startLogout: ()=>dispatch(startLogout())
    }
}

export default connect(undefined, mapDispatchToProps)(Header)