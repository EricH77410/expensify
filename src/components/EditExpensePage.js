import React from 'react'
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpense } from '../actions/expenses'

import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
    onSubmit = (e) => {
        this.props.startEditExpense(this.props.expense.id,e)
        this.props.history.push('/');
    }
    onRemove = () => {
        this.props.startRemoveExpense({id:this.props.expense.id})
        this.props.history.push('/')
    }
    render(){
        return (
            <div>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        )
    }
    
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((e)=>e.id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch,props) => {
   return {
    startEditExpense: (id,expense)=>dispatch(startEditExpense(id,expense)),
    startRemoveExpense: (e)=>dispatch(startRemoveExpense(e))
   } 
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)