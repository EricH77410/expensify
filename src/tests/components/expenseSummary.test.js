import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';

test('should render ExpenseSummary correctly with 1 expense', ()=>{
    const wrapper = shallow(<ExpenseSummary expensesCount={1} expensesTotal={235}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseSummary correctly with multiple expenses', ()=>{
    const wrapper = shallow(<ExpenseSummary expensesCount={23} expensesTotal={12543600}/>);
    expect(wrapper).toMatchSnapshot();
});