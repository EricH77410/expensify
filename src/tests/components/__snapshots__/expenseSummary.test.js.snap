// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary correctly with 1 expense 1`] = `
<div>
  <h3>
    Expense Summary
  </h3>
  <p>
    Viewing 
    1
     
    expense
  </p>
  <p>
    Total : 
    $2.35
  </p>
</div>
`;

exports[`should render ExpenseSummary correctly with multiple expenses 1`] = `
<div>
  <h3>
    Expense Summary
  </h3>
  <p>
    Viewing 
    23
     
    expenses
  </p>
  <p>
    Total : 
    $125,436.00
  </p>
</div>
`;
